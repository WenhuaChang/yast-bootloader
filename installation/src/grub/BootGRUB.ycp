/**
 * File:
 *	BootGRUB.ycp
 *
 * Module:
 *	BootGRUB
 *
 * Summary:
 *	functions for bootload setup
 *
 * Authors:
 *	Anas Nashif <nashif@suse.de>
 *
 * $Id$
 *
 */
{
    module "BootGRUB";

    textdomain "bootloader";

    import "Arch";
    import "Boot";
    import "Kernel";

    /**
     * @return boolean: LBA Support
     *
     */
    global boolean lba_support	= false;

    // ----------------------------------------------------

    /**
     * local parameters, set via UI
     */

    boolean activate = false;

    boolean linear = false;

    // ----------------------------------------------------

    /**
     * DisplayLogFile
     * Displaying log files
     * @ return 'ok_help
     */
    global define DisplayLogFile( string msg_text ) ``{
	    term text = `RichText( "<p>" + msg_text + "</p>" );
	    OpenDialog(`opt ( `decorated ),
		       `VBox (`HSpacing(75),
				// heading
			      `Heading( _("Error Log File")),
			      text,	// e.g. `Richtext()
			      `PushButton( `id(`ok_help), `opt(`default), OKButtonLabel() )
			      )
		       );

	    SetFocus(`id(`ok_help) );

	    any r = UserInput();
	    CloseDialog();
	    return (r);
    };


    /**
     * WriteUpdate
     * Calling mk_initrd and lilo while update.
     * @ return void
     */
    global define void WriteUpdate()
    ``{
	// Calling mk_initrd

	if ( SCR::Execute (.target.bash, "/sbin/mk_initrd >> /var/log/y2logmk_initrd 2>> /var/log/y2logmk_initrd") != 0 )
	{
	    any retAsk = UI::AskShowLog( "mk_initrd" );
	    if ( retAsk == `yes )
	    {
		string text = "";

		text = SCR::Read (.target.string, "/var/log/y2logmk_initrd");
		DisplayLogFile( text );
	    }
	}

	// Generate grafical login

	string menu = SCR::Read (.target.string, "/boot/grub/menu.lst");
	list lines = splitstring (menu, "\n");
	lines = filter (`l, lines, ``(substring (l, 0, 7) != "gfxmenu" && substring (l, 0, 5) != "color"));
	lines = prepend (lines, "color white/green black/light-gray");
	lines = prepend (lines, sformat ("gfxmenu %1%2%3", lilo2Grub (Boot::BootPartitionDevice), prefix,  "message"));
	menu = mergestring (lines, "\n");
	SCR::Write (.target.string, "/boot/grub/menu.lst", menu);

	string grub_conf = SCR::Read (.target.string, "/etc/grub.conf");
	string command = sformat ("echo \"%1\" | grub --batch ", grub_conf);
	SCR::Execute(.target.bash, command);
    }



    /**
     * Set
     * Set LILO parameters (used in autoinst mode)
     * @ return void
     */
    global define void Set(boolean act, boolean lin, boolean lba)
    ``{
	activate = act;
	linear = lin;
	lba_support = lba;
	return;
    }

    /**
     * ProbeLBA
     *
     * probe BIOS for lba support and set global lba_support accordingly
     */

    global define void ProbeLBA()
    ``{
	list internal_bios = SCR::Read (.probe.bios);
	lba_support = internal_bios[0, "lba_support"]:false;
	return;
    }


    /**
     * AskParameters
     *
     * helper function for inst_bootloader
     * provide ui code to ask LILO specific parameters
     *
     * @returns: list of <uicode>
     */

    global define list AskParameters ()
    ``{
	return [];
    }

    /**
     * HelpParameters
     *
     * help text for the values from AskParameters ()
     *
     * @returns: string of help text
     */

    global define string HelpParameters ()
    ``{
	return "";
    }


    /**
     * HandleParameters
     *
     * helper function for inst_bootloader
     * called after UI::UserInput() returned
     *  (only if AskParameters() returned a non-empty list)
     */

    global define void HandleParameters ()
    ``{
	return;
    }


    /**
     * write the LILO bootloader
     *
     */

    global define boolean Write ()
    ``{
	import "Mode";
	import "Boot";
	import "Misc";
	import "StorageDevices";
	import "Storage";
	import "Report";

	include "ui/common_popups.ycp";
	include "lilo/routines.ycp";
	include "lilo/io.ycp";

	SCR::Execute (.target.bash, "test -f /etc/grub.conf || touch /etc/grub.conf");
	SCR::Execute (.target.bash, "test -d /boot/grub || mkdir /boot/grub;");
	SCR::Execute (.target.bash, "test -f /boot/grub/menu.lst || touch /boot/grub/menu.lst;");
	if (! Mode::update)
	{
	    SCR::Execute (.target.bash, "test -f /boot/grub/device.map && rm /boot/grub/device.map");
	}
	SCR::Execute (.target.bash, "test -f /boot/grub/device.map || echo \"quit\" | /usr/sbin/grub --device-map=/boot/grub/device.map");

        if ( Mode::update )
        {
            // update mode
            WriteUpdate();
            return true;
        }

        // check for bios_id on a ide/scsi mixed system
	// check if bios_id is given for a target

	string dev_map = SCR::Read (.target.string, "/boot/grub/device.map");
	y2milestone ("Got device map: %1", dev_map);
	list maps = filter (`e, splitstring (dev_map, "\n"), ``(e != ""));
	maps = maplist (`e, maps, ``{return filter (`f, splitstring (e, " \t"), ``(f != ""));});
	map mapping = $[];
	foreach (`e, maps, ``{mapping = add (mapping, select (e, 1, ""), select (e, 0, ""));});

	map targetMap = Storage::GetTargetMap();
	foreach (`target_dev, `target, targetMap, ``{
	    string bios_id = target["bios_id"]:"";
	    if (bios_id != "")
	    {
	        string grub_dev = sformat ("(hd%1)", (tointeger (bios_id) - tointeger ("0x80")));
	        foreach (`e, `f, mapping, ``{
		    if (e == target_dev || f == grub_dev)
		        mapping = remove (mapping, e);
	        });
		mapping = add (mapping, target_dev, grub_dev);
	    }
	});
	dev_map = "";
	foreach (`k, `v, mapping, ``{
	    dev_map = dev_map + sformat("%1\t%2\n", v, k);
	});
	y2milestone ("Fixed device mapping: %1", dev_map);
	SCR::Write (.target.string, "/boot/grub/device.map", dev_map);


	// Boot setup start

	// additional message that needs to be displayed
	string confirm_boot_msg = Misc::boot_msg;

	integer doresult = 0;
	string docmd = "/usr/lib/YaST2/bin/dolilo /";

	string prefix = "";

	if (Boot::BootPartitionDevice == Boot::RootPartitionDevice)
	    prefix = "/boot/";
	else
	    prefix = "/";
	string root_dev = lilo2Grub (Boot::BootPartitionDevice);

	string boot_sect = Boot::device;
	map settings = $[
	    "root": root_dev,
	    "stage1" : sformat ("%1grub/stage1", prefix),
	    "stage2" : sformat ("%1grub/stage2", prefix),
	    "discswitch": true,
	    "device" : lilo2Grub (boot_sect),
	    "addr" : "0x8000",
	    "menu" : sformat ("%1%2grub/menu.lst", lilo2Grub (Boot::BootPartitionDevice), prefix)
	];
	saveGrubConf (settings);
	string grub_conf = getGrubConf (settings);

	string menu = "";
	menu = menu + sformat ("gfxmenu %1%2%3\n", lilo2Grub (Boot::BootPartitionDevice), prefix,  "message");
	menu = menu + "color white/green black/light-gray\n";
	menu = menu + "default 0\n";
	menu = menu + "timeout 8\n";

	menu = menu + "\n";

        // if syslinux recognized a framebuffer, pass it to lilo

        integer vga = tointeger (Boot::VgaType);
	string vgastring = "";
        if (vga > 0)
        {
            list fbdev = SCR::Read(.probe.fbdev);
            string unique_key = fbdev[0, "unique_key"]:"";
            if (size (unique_key) > 0)
            {
                HwStatus::Set (unique_key, `yes);
            }
	    vgastring = sformat (" vga=%1", vga);
        }

	menu = menu + "title linux\n";
	menu = menu + sformat ("   kernel %1 %2 %3\n", sformat ("%1%2%3", lilo2Grub (Boot::BootPartitionDevice), prefix,  "vmlinuz"),
	    sformat ("root=%1", Boot::RootPartitionDevice), sformat ("%1 %2", Kernel::cmdLine, vgastring));
	menu = menu + sformat ("   initrd %1%2%3\n", lilo2Grub (Boot::BootPartitionDevice), prefix,  "initrd");

	y2debug ("others = %1", Boot::others);
	if (size (Boot::others) > 0)
        {
	    foreach (`o, Boot::others, ``{
		list parts = splitstring (o, " ");
		while (select (parts, 0, " ") == "")
		    parts = remove (parts, 0);
		string dev = select (parts, 0, "");
		dev = lilo2Grub (dev);
		parts = remove (parts, 0);
		string label = mergestring (parts, " ");
		if (dev != "" && label != "")
		{
		    menu = menu + sformat ("title %1\n", label);
		    menu = menu + sformat ("   root %1\n", dev);
		    menu = menu + "   makeactive\n   chainloader +1\n";
		}
	    });
	}

	if (StorageDevices::FloppyPresent)
	{
	    menu = menu + "title floppy\n";
	    menu = menu + sformat ("   root %1\n   chainloader +1\n",
		lilo2Grub (StorageDevices::FloppyDevice));
	}

	menu = menu + "title failsafe\n";
	menu = menu + sformat ("   kernel %1 %2 %3\n", sformat ("%1%2%3", lilo2Grub (Boot::BootPartitionDevice), prefix,  "vmlinuz.shipped"),
            sformat ("root=%1", Boot::RootPartitionDevice), "ide=nodma apm=off acpi=off vga=normal nosmp maxcpus=0 disableapic 3");
	menu = menu + sformat ("   initrd %1%2%3\n", lilo2Grub (Boot::BootPartitionDevice), prefix,  "initrd.shipped");

	if (-1 != SCR::Read (.target.size, "/boot/memtest.bin"))
        {
	    menu = menu + "title memtest86\n";
	    menu = menu + sformat ("   kernel %1%2%3\n", lilo2Grub (Boot::BootPartitionDevice), prefix,  "memtest.bin");
        }

	SCR::Write (.target.string, "/boot/grub/menu.lst", menu);


	y2milestone ("Starting mk_initrd");
	SCR::Execute(.target.bash, "/sbin/mk_initrd");
	y2milestone ("Starting GRUB");
	string command = sformat ("echo \"%1\" | /usr/sbin/grub --batch ", grub_conf);
	SCR::Execute(.target.bash, command);
	y2milestone ("GRUB installed");

	return true;
    }

    /**
     *
     * Save parameters to sysconfig
     */

    global define void Save()
    ``{
    }

    /**
     * constructor
     *
     * initializes lba_support
     */

    global define void BootGRUB()
    ``{
	ProbeLBA();

	return;
    }


}
