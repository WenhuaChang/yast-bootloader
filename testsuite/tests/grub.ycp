/**
 * File:
 *  grub.ycp
 *
 * Module:
 *  Bootloader configurator
 *
 * Summary:
 *  Bootloader GRUB misc functions testsuite
 *
 * Authors:
 *  Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 */

{
// testedfiles: BootGRUB.ycp bootloader/grub/misc.ycp

    include "testsuite.ycp";

    map READ = $[
        "target" : $[
            "size" : -1,
            "tmpdir" : "/tmp/",
	    "yast2" : nil,
        ],
        "probe" : $[
            "architecture" : "i386",
            "has_apm" : true,
            "has_pcmcia" : false,
            "has_smp" : false,
            "system" : [],
            "memory" : [],
            "cpu" : [],
            "cdrom" : $[
                "manual" : [],
            ],
            "floppy" : $[
                "manual" : [],
            ],
	    "disk" : [],
	    "is_uml" : false,
        ],
        "sysconfig" : $[
            "console" : $[
                "CONSOLE_FONT" : "",
                "CONSOLE_SCREENMAP" : "",
                "CONSOLE_UNICODEMAP" : "",
                "CONSOLE_MAGIC" : "",
                "CONSOLE_ENCODING" : "",
            ],
            "language" : $[
                "RC_LANG" : "",
                "DEFAULT_LANGUAGE" : "",
            ],
        ],
        "etc" : $[
            "install_inf" : $[
                "Cmdline" : "",
            ],
        ],
        "proc" : $[
            "cpuinfo" : $[
                "value" : $[
                    "0" : $[
                        "flags" : ""
                    ],
                ],
            ],
	    "partitions" : [],
        ],
	"product" : $[
	    "features" : $[
		"USE_DESKTOP_SCHEDULER" : "0",
		"ENABLE_AUTOLOGIN" : "0",
		"EVMS_CONFIG" : "0",
		"IO_SCHEDULER" : "cfg",
		"UI_MODE" : "expert",
	    ]
	],
    ];
    map WRITE = $[];
    map EXEC = $[
	"target" : $[
	    "bash_output" : $[],
	],
    ];

    TESTSUITE_INIT ([READ, WRITE, EXEC], 0);
    import "BootGRUB";
    import "Mode";

    Mode::test = true;

    DUMP ("======================================");

    TEST (``(BootGRUB::string2devMap ("(hd0) /dev/hda\n(hd1)	/dev/hdb")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::string2devMap ("(hd1) /dev/hda\n(hd1)    /dev/hdb")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::string2devMap ("(hd0)/dev/hda\n(hd1)    /dev/hdb")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::string2devMap ("(hd1) /dev/hda\n(hd0)    /dev/hdb")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::string2devMap ("(hd1) /dev/hda")), [], 0);


    DUMP ("======================================");

    TEST (``(mergestring (splitstring (BootGRUB::devMap2string ([["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"]]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::devMap2string ([["(hd0)", "/dev/hda"], ["(hd0)", "/dev/hdb"]]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::devMap2string ([["(hd0)", "/dev/hda"]]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::devMap2string ([["(hd0)", "/dev/hda"], ["(hd1)"]]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    BootGRUB::device_map = [["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"], ["(hd2)", "/dev/cciss/c0d0"]];

    TEST (``(BootGRUB::grubDev2unixDev ("(hd0)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd1)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd2)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd0,0)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd1,4)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd2,2)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd3)")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::grubDev2unixDev ("(hd3,3)")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hda")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hdb")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/cciss/c0d0")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hda1")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hdb5")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/cciss/c0d0p3")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hdc")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::unixDev2grubDev ("/dev/hdc3")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    BootGRUB::device_map = [["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"], ["(hd2)", "/dev/cciss/c0d0"], ["(fd0)", "/dev/fd0" ]];

    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  --stage2=/boot/grub/stage2 /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  --stage2=/boot/grub/stage2 /grub/stage1 (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  /grub/stage1 (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  --stage2=/boot/grub/stage2 --force-lba=off /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    BootGRUB::device_map = [["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"], ["(hd2)", "/dev/cciss/c0d0"], ["(fd0)", "/dev/fd0" ]];

    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"(fd0)", "discswitch":"d", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "prefix":" --stage2=/boot/grub/stage2", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"(fd0)", "discswitch":"d", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"(fd0)", "discswitch":"", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "prefix":" --stage2=/boot/grub/stage2", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"(fd0)", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"(fd0)", "discswitch":"d", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "prefix":" --stage2=/boot/grub/stage2 --force-lba=off", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);
    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($[]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd1)aaa"]], $["(hd0)" : "(hd1)", "(hd1)" : "(hd0)"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd1)aaa"]], $["(hd0)" : "(hd1)", "(hd1)" : "(hd1)"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd1)aaa"]], $["(hd0)" : "(hd1)"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd1)aaa"]], $["(hd0)" : "(hd0)", "(hd1)" : "(hd1)"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd2)aaa"]], $["(hd0)" : "(hd1)", "(hd1)" : "(hd0)"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::remapDisks ([$["key" : "A", "value" : "(hd0)(hd1)"], $[ "key" : "B", "value" : "aaa(hd2)aaa"]], $["(hd0)" : "(hd1)", "(hd1)" : "(hd2)", "(hd2)" : "(hd0)"])), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    TEST (``(BootGRUB::splitDevPath ("(hd0,0)/boot/vmlinuz")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::splitDevPath ("(hd0,0)/vmlinuz")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::splitDevPath ("/boot/vmlinuz")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::splitDevPath ("(hd0,0)")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    TEST (``(BootGRUB::mergeDevPath (["(hd0,0)", "/boot/vmlinuz"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::mergeDevPath (["(hd0,0)", "boot/vmlinuz"])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::mergeDevPath (["(hd0,0)", ""])), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::mergeDevPath ([nil, "/boot/vmlinuz"])), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    BootGRUB::device_map = [["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"], ["(hd2)", "/dev/cciss/c0d0"]];

    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"/dev/md0", "discswitch":"d", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "prefix":" --stage2=/boot/grub/stage2 --force-lba=off", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);

    TEST (``(mergestring (splitstring (BootGRUB::getGrubConf ($["addr":"0x8000", "device":"mbr_md", "discswitch":"d", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "prefix":" --stage2=/boot/grub/stage2 --force-lba=off", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]), "\n"), "\\n")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");

    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  /grub/stage1 (hd0,0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
install  /grub/stage1 (hd1,0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);
    TEST (``(BootGRUB::parseGrubConf ("root (hd0,0)
install  /grub/stage1 (hd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
install  /grub/stage1 (hd1) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst
quit")), [READ, WRITE, EXEC], 0);

    DUMP ("======================================");
}
