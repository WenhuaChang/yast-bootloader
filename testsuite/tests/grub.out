Dump	======================================
Return	[["(hd0)", "/dev/hda"], ["(hd1)", "/dev/hdb"]]
Return	[["(hd1)", "/dev/hda"], ["(hd1)", "/dev/hdb"]]
Return	[["(hd0)/dev/hda"], ["(hd1)", "/dev/hdb"]]
Return	[["(hd1)", "/dev/hda"], ["(hd0)", "/dev/hdb"]]
Return	[["(hd1)", "/dev/hda"]]
Dump	======================================
Return	(hd0)	/dev/hda\n(hd1)	/dev/hdb\n
Return	(hd0)	/dev/hda\n(hd0)	/dev/hdb\n
Return	(hd0)	/dev/hda\n
Return	(hd0)	/dev/hda\n(hd1)	\n
Dump	======================================
Return	/dev/hda
Return	/dev/hdb
Return	/dev/cciss/c0d0
Return	/dev/hda1
Return	/dev/hdb5
Return	/dev/cciss/c0d0p3
Return	/dev/null
Return	/dev/null
Dump	======================================
Return	(hd0)
Return	(hd1)
Return	(hd2)
Return	(hd0,0)
Return	(hd1,4)
Return	(hd2,2)
Return	(null)
Return	(null)
Dump	======================================
Return	$["addr":"0x8000", "device":"(fd0)", "discswitch":true, "menu":"(hd0,0)/grub/menu.lst", "prefix":" --stage2=/boot/grub/stage2", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]
Return	$["addr":"0x8000", "device":"(fd0)", "discswitch":true, "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]
Return	$["addr":"0x8000", "device":"(fd0)", "menu":"(hd0,0)/grub/menu.lst", "prefix":" --stage2=/boot/grub/stage2", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]
Return	$["addr":"0x8000", "device":"(fd0)", "menu":"(hd0,0)/grub/menu.lst", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]
Return	$["addr":"0x8000", "device":"(fd0)", "discswitch":true, "menu":"(hd0,0)/grub/menu.lst", "prefix":" --stage2=/boot/grub/stage2 --force-lba=off", "root":"(hd0,0)", "stage1":"/grub/stage1", "stage2":"/grub/stage2"]
Return	$[]
Dump	======================================
Return	root (hd0,0)\ninstall  --stage2=/boot/grub/stage2 /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst\nquit\n
Return	root (hd0,0)\ninstall  /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst\nquit\n
Return	root (hd0,0)\ninstall  --stage2=/boot/grub/stage2 /grub/stage1 (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst\nquit\n
Return	root (hd0,0)\ninstall  /grub/stage1 (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst\nquit\n
Return	root (hd0,0)\ninstall  --stage2=/boot/grub/stage2 --force-lba=off /grub/stage1 d (fd0) /grub/stage2 0x8000 (hd0,0)/grub/menu.lst\nquit\n
Return	root \ninstall      \nquit\n
Dump	======================================
Return	[$["key":"A", "value":"(hd1)(hd0)"], $["key":"B", "value":"aaa(hd0)aaa"]]
Return	[$["key":"A", "value":"(hd1)(hd1)"], $["key":"B", "value":"aaa(hd1)aaa"]]
Return	[$["key":"A", "value":"(hd1)(hd1)"], $["key":"B", "value":"aaa(hd1)aaa"]]
Return	[$["key":"A", "value":"(hd0)(hd1)"], $["key":"B", "value":"aaa(hd1)aaa"]]
Return	[$["key":"A", "value":"(hd1)(hd0)"], $["key":"B", "value":"aaa(hd2)aaa"]]
Return	[$["key":"A", "value":"(hd1)(hd2)"], $["key":"B", "value":"aaa(hd0)aaa"]]
Dump	======================================
Return	["(hd0,0)", "/boot/vmlinuz"]
Return	["(hd0,0)", "/vmlinuz"]
Return	[nil, "/boot/vmlinuz"]
Return	["(hd0,0)", ""]
Dump	======================================
Return	(hd0,0)/boot/vmlinuz
Return	(hd0,0)/boot/vmlinuz
Return	(hd0,0)/
Return	nil
Dump	======================================
